src/components/Identity/injectEnsureRequestHasIdentity.js

```diff
@@ -18,38 +18,72 @@ export default ({
    * we queue all other requests, otherwise the requests could result in
-   * multiple ECIDs being minted for the user. Once the response to the first
-   * request returns, we can let the queued requests be sent, since they
-   * will have the newly minted ECID that was returned on the first response.
+   * multiple ECIDs being minted for the user. Once we get an identity
+   * cookie, we can let the queued requests be sent all at once, since they
+   * will have the newly minted ECID.
+   *
+   * Konductor should make every effort to return an identity, but in
+   * certain scenarios it may not. For example, in cases where the
+   * request does not match what Konductor is expecting (ie 400s).
+   * In cases where Konductor does not set an identity, there should be
+   * no events recorded so we don't need to worry about multiple ECIDs
+   * being minted for each user.
+   *
+   * If we only allowed one request through without an identity, a single
+   * malformed request causes all other requests to never send.
    */
   return ({ payload, onResponse }) => {
     if (doesIdentityCookieExist()) {
       return Promise.resolve();
     }
 
-    if (identityCookiePromise) {
-      // We don't have an identity cookie, but the first request has
-      // been sent to get it. We must wait for the response to the first
-      // request to come back and a cookie set before we can let this
-      // request go out.
+    if (obtainedIdentityPromise) {
+      // We don't have an identity cookie, but at least one request has
+      // been sent to get it. Konductor may set the identity cookie in the
+      // response. We will hold up this request until the last request
+      // requiring identity returns and awaitIdentityCookie confirms the
+      // identity was set.
       logger.log("Delaying request while retrieving ECID from server.");
-      return identityCookiePromise.then(() => {
-        logger.log("Resuming previously delayed request.");
+      const previousObtainedIdentityPromise = obtainedIdentityPromise;
+
+      // This promise resolves when we have an identity cookie. Additional
+      // requests are chained together so that only one is sent at a time
+      // until we have the identity cookie.
+      obtainedIdentityPromise = previousObtainedIdentityPromise.catch(() => {
+        return awaitIdentityCookie(onResponse);
       });
+
+      // When this returned promise resolves, the request will go out.
+      return (
+        previousObtainedIdentityPromise
+          .then(() => {
+            logger.log("Resuming previously delayed request.");
+          })
+          // If Konductor did not set the identity cookie on the previous
+          // request, then awaitIdentityCookie will reject its promise.
+          // Catch the rejection here and allow this request to go out.
+          .catch(() => {
+            return allowRequestToGoWithoutIdentity(payload);
```

Comment:
What about the requests that doesn't result in errors?

Reply:
Requests that don't result in errors would trigger the then clause, and so the promise would resolve after logging. Once the promise resolves, the request would go out. Is there a particular use case you are thinking about?

Reply:
But does it get sent (this would be request 2,3,etc) if the first req doesn't end in error? What am I missing?

Reply:
Yes it gets sent. The Identity component calls ensureRequestHasIdentity in the "onBeforeRequest" lifecycle method and returns the promise.  This lifecycle method waits for all the promises returned from "onBeforeRequest" to resolve before sending the request. So when the first request comes back with the identity, the promise resolves and the others are able to go out.

Reply:
👍